// @theme-ui/preset-funk/dist/theme-ui-preset-funk.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var base = require('@theme-ui/preset-base');
var utils = require('@theme-ui/css/utils');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var base__default = /*#__PURE__*/_interopDefault(base);

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

var funk = utils.makeTheme(_extends({}, base__default['default'], {
  fonts: {
    body: 'Poppins, sans-serif',
    heading: 'Poppins, sans-serif',
    monospace: 'Menlo, monospace'
  },
  lineHeights: {
    body: 1.625,
    heading: 1.25
  },
  fontWeights: {
    body: 400,
    heading: 900,
    bold: 700
  },
  colors: _extends({}, base__default['default'].colors, {
    primary: '#609',
    secondary: '#306'
  }),
  styles: _extends({}, base__default['default'].styles)
}));

exports.default = funk;
exports.funk = funk;
