// @theme-ui/presets/dist/theme-ui-presets.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var presetBase = require('@theme-ui/preset-base');
var presetDark = require('@theme-ui/preset-dark');
var presetDeep = require('@theme-ui/preset-deep');
var presetFunk = require('@theme-ui/preset-funk');
var presetFuture = require('@theme-ui/preset-future');
var presetRoboto = require('@theme-ui/preset-roboto');
var presetSwiss = require('@theme-ui/preset-swiss');
var presetSystem = require('@theme-ui/preset-system');
var presetTosh = require('@theme-ui/preset-tosh');
var presetBootstrap = require('@theme-ui/preset-bootstrap');
var presetBulma = require('@theme-ui/preset-bulma');
var presetPolaris = require('@theme-ui/preset-polaris');
var presetTailwind = require('@theme-ui/preset-tailwind');
var presetSketchy = require('@theme-ui/preset-sketchy');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var presetBase__default = /*#__PURE__*/_interopDefault(presetBase);
var presetDark__default = /*#__PURE__*/_interopDefault(presetDark);
var presetDeep__default = /*#__PURE__*/_interopDefault(presetDeep);
var presetFunk__default = /*#__PURE__*/_interopDefault(presetFunk);
var presetFuture__default = /*#__PURE__*/_interopDefault(presetFuture);
var presetRoboto__default = /*#__PURE__*/_interopDefault(presetRoboto);
var presetSwiss__default = /*#__PURE__*/_interopDefault(presetSwiss);
var presetSystem__default = /*#__PURE__*/_interopDefault(presetSystem);
var presetTosh__default = /*#__PURE__*/_interopDefault(presetTosh);
var presetBootstrap__default = /*#__PURE__*/_interopDefault(presetBootstrap);
var presetBulma__default = /*#__PURE__*/_interopDefault(presetBulma);
var presetPolaris__default = /*#__PURE__*/_interopDefault(presetPolaris);
var presetTailwind__default = /*#__PURE__*/_interopDefault(presetTailwind);
var presetSketchy__default = /*#__PURE__*/_interopDefault(presetSketchy);



Object.defineProperty(exports, 'base', {
	enumerable: true,
	get: function () {
		return presetBase__default['default'];
	}
});
Object.defineProperty(exports, 'dark', {
	enumerable: true,
	get: function () {
		return presetDark__default['default'];
	}
});
Object.defineProperty(exports, 'deep', {
	enumerable: true,
	get: function () {
		return presetDeep__default['default'];
	}
});
Object.defineProperty(exports, 'funk', {
	enumerable: true,
	get: function () {
		return presetFunk__default['default'];
	}
});
Object.defineProperty(exports, 'future', {
	enumerable: true,
	get: function () {
		return presetFuture__default['default'];
	}
});
Object.defineProperty(exports, 'roboto', {
	enumerable: true,
	get: function () {
		return presetRoboto__default['default'];
	}
});
Object.defineProperty(exports, 'swiss', {
	enumerable: true,
	get: function () {
		return presetSwiss__default['default'];
	}
});
Object.defineProperty(exports, 'system', {
	enumerable: true,
	get: function () {
		return presetSystem__default['default'];
	}
});
Object.defineProperty(exports, 'tosh', {
	enumerable: true,
	get: function () {
		return presetTosh__default['default'];
	}
});
Object.defineProperty(exports, 'bootstrap', {
	enumerable: true,
	get: function () {
		return presetBootstrap__default['default'];
	}
});
Object.defineProperty(exports, 'bulma', {
	enumerable: true,
	get: function () {
		return presetBulma__default['default'];
	}
});
Object.defineProperty(exports, 'polaris', {
	enumerable: true,
	get: function () {
		return presetPolaris__default['default'];
	}
});
Object.defineProperty(exports, 'tailwind', {
	enumerable: true,
	get: function () {
		return presetTailwind__default['default'];
	}
});
Object.defineProperty(exports, 'sketchy', {
	enumerable: true,
	get: function () {
		return presetSketchy__default['default'];
	}
});
